src/uct/api/uct.h

```diff
@@ -1267,19 +1267,20 @@ typedef struct uct_md_mem_attr {
  * @brief Query attributes of a given pointer
  *
  * Return attributes such as memory type, and system device for the
- * given pointer of specific length.
+ * given pointer of specific length depending on fields requested in mem_attr.
  *
- * @param [in]     md          Memory domain to run the query on. This function
- *                             returns an error if the md does not recognize the
- *                             pointer.
+ * @param [in]     md          Memory domain to run the query on.
  * @param [in]     address     The address of the pointer. Must be non-NULL
  *                             else UCS_ERR_INVALID_PARAM error is returned.
  * @param [in]     length      Length of the memory region to examine.
  *                             Must be nonzero else UCS_ERR_INVALID_PARAM error
  *                             is returned.
- * @param [out]    mem_attr    If successful, filled with ptr attributes.
+ * @param [in,out] mem_attr    MD attribute param specifying fields of interest.
+ *                             If successful, filled with ptr attributes.
  *
- * @return Error code.
+ * @return This function returns an appropriate error code if the md does not
+ *         recognize the pointer or if all of the requested fields in mem_attr
+ *         cannot be populated.
```

Comment:
how does the caller of this function know what fields exactly it can query? for example, Rocm can't query UCT_MD_MEM_ATTR_FIELD_SYS_DEV

Reply:
all UCTs should implement filling all information, we can define some "default" values if needed , like "UNKNOWN"

Reply:
@Akshay-Venkatesh need to fix rocm side by setting UNKNOWN instead of failing for SYS_DEV

Reply:
@yosefe Started looking at defining what an unknown `ucs_sys_device_t` is but we may have an issue:

```
typedef struct ucs_sys_bus_id {
    uint16_t domain;   /* range: 0 to ffff */
    uint8_t  bus;      /* range: 0 to ff */
    uint8_t  slot;     /* range: 0 to 1f */
    uint8_t  function; /* range: 0 to 7 */
} ucs_sys_bus_id_t;**
 
typedef struct ucs_sys_device {
    unsigned         id;        /**< Index of the device */
    int              numa_node; /**< NUMA node assoicated with the device*/
    ucs_sys_bus_id_t bus_id;    /**< bus ID of of the device if applicable.
                                   eg: 0000:06:00.0 {domain:bus:slot.function}*/
} ucs_sys_device_t;
```

If we decide to set all fields in `ucs_sys_device_t` to 0 (as the definition of unknown sys dev), then this potentially ends up pointing to PCIe root complex (0000:00:00.0). One way to handle this is to introduce a member `sys_type` in  `ucs_sys_device_t` which is a bit_mask again:

```
typedef enum ucs_sys_dev_type {
UCS_SYS_DEV_NET,
UCS_SYS_DEV_MEM,
UCS_SYS_DEV_UNKNOWN
} ucs_sys_dev_type_t;

typedef struct ucs_sys_device {
    unsigned         id;        /**< Index of the device */
    int              numa_node; /**< NUMA node assoicated with the device*/
    ucs_sys_bus_id_t bus_id;    /**< bus ID of of the device if applicable.
                                   eg: 0000:06:00.0 {domain:bus:slot.function}*/
   uint64_t     sys_type; /* UCS_SYS_DEV_UNKNOWN by default else an or'ing of
                                 UCS_SYS_DEV_NET, UCS_SYS_DEV_MEM,*/
} ucs_sys_device_t;
```

Once we add this field, we can then check if sys device type is unknown. Does that work?